File: Нейронные сети/lpc_nn/source/lpc55-C/random.c
/***********************************************************************
*
*	RANDOM Version 49
*
***********************************************************************
*
*  Pseudo random number generator based on Knuth, Vol 2, p. 27.
*
* Function Return:
*  RANDOM - Integer variable, uniformly distributed over -32768 to 32767
*
* Warning:  This routine contains statements which are specific to VAX Fortran.
*		I'll try it anyway - Tiner
*/

#define MIDTAP 1
#define MAXTAP 4

int Rrandom ()
{
int the_random;
static short y[MAXTAP+1]={-21161, -8478, 30892,-10216, 16950};
static int j=MIDTAP, k=MAXTAP;

/*   The following is a 16 bit 2's complement addition,
*   with overflow checking disabled	*/

y[k] += y[j];

if(y[k] > 32767) /* to make 16 bit rollover to -/+ */
  y[k] = -(32768 - (y[k] & 32767));
if(y[k] < -32768)
  y[k] = y[k] & 32767;
	
the_random = y[k];
k--;
if (k < 0) k = MAXTAP;
j--;
if (j < 0) j = MAXTAP;

return(the_random);

}

File: Нейронные сети/lpc_nn/source/lpc55/random.f
***********************************************************************
*
*	RANDOM Version 49
*
**********************************************************************
*
*  Pseudo random number generator based on Knuth, Vol 2, p. 27.
*
* Function Return:
*  RANDOM - Integer variable, uniformly distributed over -32768 to 32767
*
	FUNCTION RANDOM ()
	INTEGER RANDOM, J, K, MIDTAP, MAXTAP
	PARAMETER (MIDTAP=2, MAXTAP=5)
	INTEGER*2 Y(MAXTAP)
	SAVE J, K, Y
	DATA Y /-21161, -8478, 30892,-10216, 16950/
	DATA J/MIDTAP/, K/MAXTAP/

*   The following is a 16 bit 2's complement addition,
*   with overflow checking disabled

	Y(K) = Y(K) + Y(J)
	RANDOM = Y(K)
	K = K - 1
	IF (K .LE. 0) K = MAXTAP
	J = J - 1
	IF (J .LE. 0) J = MAXTAP

	RETURN
	END
